airflow/providers/openlineage/extractors/python.py

```diff
@@ -0,0 +1,92 @@
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import inspect
+import os
+from typing import Callable
+
+from airflow.providers.openlineage.extractors.base import BaseExtractor, OperatorLineage
+from airflow.providers.openlineage.plugins.facets import (
+    UnknownOperatorAttributeRunFacet,
+    UnknownOperatorInstance,
+)
+from openlineage.client.facet import SourceCodeJobFacet
+
+"""
+:meta private:
+"""
+
+
+class PythonExtractor(BaseExtractor):
+    """
+    This extractor provides visibility on what particular task does by extracting
+    executed source code and putting it into SourceCodeJobFacet. It does not extract
+    datasets.
+
+    :meta private:
+    """
+
+    @classmethod
+    def get_operator_classnames(cls) -> list[str]:
+        return ["PythonOperator"]
+
+    def extract(self) -> OperatorLineage | None:
+        # TODO Airflow config
+        collect_source = os.environ.get("OPENLINEAGE_AIRFLOW_DISABLE_SOURCE_CODE", "True").lower() not in (
+            "true",
+            "1",
+            "t",
+        )
+
+        source_code = self.get_source_code(self.operator.python_callable)
+        job_facet: dict = {}
+        if collect_source and source_code:
+            job_facet = {
+                "sourceCode": SourceCodeJobFacet(
+                    language="python",
+                    # We're on worker and should have access to DAG files
+                    source=source_code,
+                )
+            }
+        return OperatorLineage(
+            job_facets=job_facet,
+            run_facets={
+                # The BashOperator is recorded as an "unknownSource" even though we have an
+                # extractor, as the <i>data lineage</i> cannot be determined from the operator
+                # directly.
+                "unknownSourceAttribute": UnknownOperatorAttributeRunFacet(
+                    unknownItems=[
+                        UnknownOperatorInstance(
+                            name="PythonOperator",
+                            properties={attr: value for attr, value in self.operator.__dict__.items()},
+                        )
+                    ]
+                )
+            },
+        )
+
+    def get_source_code(self, callable: Callable) -> str | None:
+        try:
+            return inspect.getsource(callable)
+        except TypeError:
+            # Trying to extract source code of builtin_function_or_method
+            return str(callable)
+        except OSError:
+            self.log.exception(f"Can't get source code facet of PythonOperator {self.operator.task_id}")
```

Comment:
```suggestion
            self.log.exception("Can't get source code facet of PythonOperator %s", self.operator.task_id)
```
To keep with the consistency of using %-formatting for logging.

Reply:
@josh-fell done